### ISSUE TITLE
Implement .bulk_update(...)

### ISSUE THREAD

**Issue by hynekcer** [django-salesforce/django-salesforce#236]
The method [bulk_update](https://docs.djangoproject.com/en/2.2/ref/models/querysets/#s-bulk-update) is new in Django 2.2.

Implement it in Django-salesforce.

Notes: A method `salesforce.dbapi.driver.RawConnection.sobject_collections_request(...)` was written for this purpose. The implementation in Django is completely different. Therefore we will not use any code from a Django method and easily backport it for Salesforce backend also to an old Django.

**Comment by garrypolley** [comment]
Is there active work on this issue taking place in a different branch/fork? I'm thinking of potentially looking into helping out here. We have some API limit issues with our updates and batching the updates in groups via the bulk_update would help us out a lot.

**Comment by hynekcer** [comment]
Thanks for your interest and the offer. I can write some useful details: I think that the driver part with `cursor.db.connection.sobject_collections_request('PATCH', records)`  has been worked correctly since v0.8.1. We should (and you can start it until you can) monkey paste the method `bulk_update` ([source code](https://github.com/django/django/blob/efc1c73bf511cbebf03fa5b13fa3a98718daf928/django/db/models/query.py#L517-L559)) to `salesforce.backend.query.SalesforceQuerySet`. You see there a similar `bulk_create`
 The problematic useless part in the original `bulk_update` is with CASE ... WHEN ... is [here](https://github.com/django/django/blob/efc1c73bf511cbebf03fa5b13fa3a98718daf928/django/db/models/query.py#L548-L550). Therefore we will not call it by `super()` at all and useful parts like checks should be copied to our method.

Also tests are useful also for understanding the SFDC API. I have test only for `bulk_create` in `tests.test_mock.test_data.test_create()` (I use "mocked" simplified recorded real responses of new SFDC API features that were documented with no or little examples because sometimes later after an new SFDC version it is easier for maintenance (if a problem is not found while in the release preview window). I think that e. g. error reporting with or without "allOrNone" has been changed in some `/services/data/{API_VERSION}/composite/sobjects` and the change can not be verified because if even doesn't depend on API_VERSION parameter but only the release.)

It is still not clear what is the preferred way for django-salesforce users with handling of `allOrNone` parameter - that is similar to turn on transactions instead of autocommit. The batch size is 200 records for Salesforce. It is not possible to create a bigger "pseudo-transaction" than 200 records. It was also problematic, how to report succeeded / failed / cancelled updates to the end user if it is a mix (for records > 200). Maybe add a parameter `all_or_none` to the method. Some people use SalesforceModels on sqlite for simple fast tests.

So the minimal implementation is not too complicated, but there are many details and the final solution will require a discussion. Please can you start some work?

**Comment by archaean** [comment]
I have interest in this too, and will hopefully contribute with @garrypolley (@garrypolley and I are colleagues).  @hynekcer your background on the issue is very help and highlights some of my current questions about best practices regarding the use of the django-salesforce package.  Already, due to performance reasons, I have already started to leverage the `bulk_create` (e.g. single creation through was taking 1 to 2 seconds per object, where as `bulk_create` could **atomically** make 200 in ~30 seconds).

However, when using `bulk_create`, I found that I needed to fully aware of the 200 batch size limit before trying to use it, otherwise I could get into a weird state between the model and the data pushed to salesforce.  If, for example, I had 600 SF objects that I was trying to `bulk_create`, which would be 3 batches ideally, and in the second batch one of the objects would fail to create (due to a SF constraint error or there was just a general timeout) I wouldn't know, programmatically, given the state of the model, how to get the remain creations through.

I worked around this problem by explicitly chunking my model into 200 size chunks, then trying to do an `all_or_none`,  and if one of the items failed, I would know explicitly that none of the records made it into Salesforce.  From there I could try each model in the batch serially and log the specific ones that failed and continue unobstructed through the remainder of the batches.

I think this pattern would work, more or less, the same for  `bulk_update` given the "pseudo-transaction" problem.  That is when dealing with Salesforce through this package thinking in terms of atomic database transactions with rollback features is quite dangerous, as noted in the README.

I may go as far to suggest that using `bulk_create` or `bulk_update` with a Queryset larger than 200 should, by default, throw an error or issue a strong warning suggesting chunking.

**Comment by hynekcer** [comment]
I use an iterator function `chunked(iterable, n)` (below) and splitting can be written briefly with it. In the old versions of Django >= 1.4, < 2.0 it was not declared in docs that objects are casted to a list if the bulk_create parameter is a generator (and that it should be never a huge amount that can not fin in the memory). With a parameter that fit into a list it is possible to save the lists of succeeded / failed / cancelled objects to some ephemeral internal variable. (Useful e.g. to put it as a json blob to some normal database. A list of 199 primary keys by 15/18 characters is too much for a line of log that can be truncated). It is probably better that error handling in the half of a loop and it can be fixed consistently. If django-salesforce is used for data migration then pseudo-atomicity is important but in a web or mobile application it is probably preferable over allOrNone to minimize data loss, even if the orphan data should remain in some inactive state for fixing.

I know a call that allows to get not only primary keys, but also some auto updated fields like `Owner` or fields with DEFAULTED_ON_CREATE. It is also possible to read the objects before writing in the same transaction. It decreases the number of available objects written  by one request. I didn't tried it. A better implementation of Django integration with the current `sobject_collections_request` is a more basic important task.

A bulk_create with 200 simple objects is possible in 4 seconds. The time also depends on size of data, triggers, validation rules etc.

Some bulk create / update can be written also as a normal function because objects of different types can be written by the same request.

---

``` python
from itertools import islice

def chunked(iterable, n):
    """
    Break an iterable into lists of a given length::

    >>> for chunk in chunked([1, 2, 3, 4, 5], 3):
    ...     print(chunk)
    [1, 2, 3]
    [4, 5]
    """
    iterable = iter(iterable)
    while True:
        chunk = list(islice(iterable, n))
        if not chunk:
            return
        yield chunk
```
```

**Comment by hynekcer** [comment]
A simple function `bulk_update` for max 200 objects is the branch [hynekcer/bulk_update_small](https://github.com/hynekcer/django-salesforce/tree/bulk_update_small). The updated objects can be from mixed models. Maximum of 10 groups of identical models are possible. e.g a list [Account, Contact, Contact, Account] are three groups.

The next step will be to prepare a more complete state information after error. An internal loop by chunks would be not useful without it.

**Comment by dhbahr** [comment]
@hynekcer is this already merged? at the moment the `bulk_update` option is failing to serialize some syntax tree node for the field to be updated:

`models.CustomSalesforceObject.objects.bulk_update(toUpdate, ['fieldName'])`

is resulting in

`TypeError: Object of type Case is not JSON serializable`

The `kwargs` variable at the moment of the `session.request` in `salesforce.dbapi.driver` has this payload:

```{
  'json': {
    'allOrNone': True,
    'records': [
      {
        'FieldName__c': <Case: CASE WHEN <SalesforceWhereNode: (AND: <django.db.models.lookups.Exact object at 0x7f999a1312e0>)> THEN Value(Available), WHEN <SalesforceWhereNode: (AND: <django.db.models.lookups.Exact object at 0x7f999a131910>)> THEN Value(Available), ELSE Value(None)>,
        'attributes': {'type': 'FieldName__c'},
        'id': 'a0QxxxxxxxxxxxxUAS'
      }
    ]
  }
}
```

Is this something your implementation would solve?

Best

**Comment by hynekcer** [comment]
The method `.bulk_update()` is implemented by the commit [aac4cfb](https://github.com/django-salesforce/django-salesforce/commit/aac4cfbe8f35f99d5711290c76321b19c2cf7132)

`def bulk_update(self, objs, fields,  batch_size=None, all_or_none=None)`

The list `objs` is processed by batches of size `batch_size`.
The loop is terminated by any error in a batch. If the parameter `all_or_none=True` is set then no record is updated by the batch where an error occurs. 
The default value of `batch_size` is also the maximal value 200. Higher values are reduced.

A multi-line error message is written after errors, with error summary and individual lines for every error in a batch.

`    Error Summary: errors=2, rollback/cancel=199, success=0`

Possible maximum is 200 errors.

**Comment by hynekcer** [comment]
The queryset methods `.bulk_create(...)`, `.update(...)`, `.delete(...)` can be run with option `all_or_none` by the method `.sf(all_or_none=True)` added somewhere before them in the queryset chain of a SalesforceModel.

``` python
Contact.objects.sf(all_or_none=True).bulk_create(...)
```
That works also with normal databases, where the option the option `all_or_none` is ignored. If a method `.sf()` should exist in querysets for non-salesforce databases (e.g in tests) then the `SalesforceModel` should be imported from module `salesforce.models_extend`. That will be maybe a normal SalesforceModel, but currently it must be enabled explicitly by user that want it to can send a feedback continue with the old class until the new become generally acceptable.
